replit_river/codegen/client.py

```diff
@@ -49,6 +53,59 @@ def encode_type(
     if isinstance(type, RiverNotType):
         return ("None", ())
     if isinstance(type, RiverUnionType):
+        # First check if it's a discriminated union.
+        if len(type.anyOf) > 0 and all(
+            isinstance(t, RiverConcreteType)
+            and t.type == "object"
+            and t.properties
+            and (not t.patternProperties or "^(.*)$" not in t.patternProperties)
+            for t in type.anyOf
+        ):
+            # We have established that it is a union-of-objects. Now let's see if
+            # there is a discriminator field common among all options.
+            literal_fields = set[str]()
+            for i, t in enumerate(type.anyOf):
+                # Mypy already forgot this fact D:
+                assert isinstance(t, RiverConcreteType)
+                lf = set[str](
+                    name
+                    for name, prop in t.properties.items()
+                    if isinstance(prop, RiverConcreteType)
+                    and prop.type in ("string", "number", "boolean")
+                    and prop.const is not None
+                )
+                if i == 0:
+                    literal_fields = lf
+                else:
+                    literal_fields.intersection_update(lf)
+                if not literal_fields:
+                    # There are no more candidates.
+                    break
+            if len(literal_fields) == 1:
+                # Hooray! we found a discriminated union.
+                discriminator_name = literal_fields.pop()
+                one_of: List[str] = []
+
+                for t in type.anyOf:
+                    # Mypy already forgot this fact D:
+                    assert isinstance(t, RiverConcreteType)
+                    discriminator_value = [
+                        _NON_ALNUM_RE.sub("", str(prop.const))
+                        for name, prop in t.properties.items()
+                        if isinstance(prop, RiverConcreteType)
+                        and name == discriminator_name
+                        and prop.const is not None
+                    ].pop()
+                    type_name, type_chunks = encode_type(
+                        t, f"{prefix}OneOf_{discriminator_value}", base_model
+                    )
+                    chunks.extend(type_chunks)
+                    one_of.append(type_name)
+                if discriminator_name == "$kind":
+                    discriminator_name = "kind"
```

Comment:
Maybe we should reconsider these dollar signs to eliminate the extra logic needed for python. Plus "type" is a more idiomatic in typescript, and we're using isinstance/classes/nominal typing in python

Reply:
it's really Python's fault for not permitting `$` as a valid identifier character.

Alternately config file.
Alternately alternately, generic heuristic to translate or remove invalid identifier characters, with the assertion that there's no name conflict introduced.

Reply:
this ^ re: Python's fault.

fortunately the only place where it really matters is when building an object literal in Python. for that case, the change in L164 will address that.